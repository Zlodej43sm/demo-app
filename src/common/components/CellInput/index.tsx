import React, {
  useState,
  useEffect,
  ComponentType,
  SetStateAction,
  Dispatch,
} from "react";
import { useDispatch, useSelector } from "react-redux";
import { Box, Input } from "@mui/material";

import { ReactComponent as EditIcon } from "../../../assets/icons/edit.svg";
import {
  selectSpreadsheetData,
  selectSpreadsheetDataById,
} from "../../../selectors";
import { AppDispatch } from "../../../redux/store";
import useStyles from "./styles";
import {
  computeCellValue,
  setProgressStatus,
  saveData,
} from "../../../redux/SpreadsheetSlice";

export type CurrentSelectedRowProp = null | number;

export type Props = {
  id: string;
  setCurrentEditRow: Dispatch<SetStateAction<CurrentSelectedRowProp>>;
};

const CellInput: ComponentType<Props> = ({ id, setCurrentEditRow }) => {
  const classes = useStyles();
  const [isEdit, setIsEdit] = useState(false);
  const dispatch = useDispatch<AppDispatch>();
  const spreadsheetDataById = useSelector(selectSpreadsheetDataById);
  const spreadsheetData = useSelector(selectSpreadsheetData);
  const [rowIndex, colIndex] = spreadsheetDataById[id];
  const { value, formula } = spreadsheetData[rowIndex][colIndex];
  const [inputValue, setInputValue] = useState(formula || value);

  useEffect(() => {
    dispatch(computeCellValue({ id, val: inputValue }));
  }, [dispatch, id, inputValue, spreadsheetData]);

  const onEditClick = () => {
    setIsEdit(true);
    setCurrentEditRow(rowIndex);
  };
  const onInputBlur = () => {
    setIsEdit(false);
    setCurrentEditRow(null);
    dispatch(setProgressStatus({ rowIndex }));
    dispatch(saveData({ id, data: spreadsheetData }));
  };
  const onInputChange = (val: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(val.currentTarget.value);
  };

  return (
    <Box
      onClick={onEditClick}
      className={
        colIndex === 2
          ? `${classes.editBoxLast} ${classes.editBoxRoot}`
          : classes.editBoxRoot
      }
    >
      {isEdit ? (
        <Input
          disableUnderline
          classes={{ root: classes.editFieldRoot }}
          onBlur={onInputBlur}
          onChange={onInputChange}
          autoFocus
          value={inputValue}
        />
      ) : (
        value || "NaN"
      )}
      <EditIcon className={classes.editIcon} />
    </Box>
  );
};

export default CellInput;
