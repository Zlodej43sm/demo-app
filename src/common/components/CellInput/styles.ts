import { makeStyles } from "@mui/styles";

export const COMPONENT_NAME = "Spreadsheet";

export default makeStyles(
  (theme: any) => ({
    editBoxRoot: {
      position: "relative",
      zIndex: 1,

      "&:not($editBoxLast)::after": {
        content: "''",
        borderRight: `1px solid ${theme.palette.grey["300"]}`,
        position: "absolute",
        opacity: 0.4,
        top: -6,
        bottom: -6,
        right: -10,
      },
    },
    editIcon: {
      position: "absolute",
      right: -2,
      bottom: -2,
    },
    editBoxLast: {},
    editFieldRoot: {
      "& input": {
        textAlign: "center",
      },
    },
  }),
  { name: COMPONENT_NAME }
);
