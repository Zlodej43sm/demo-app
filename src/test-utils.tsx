import React, { ComponentType, PropsWithChildren } from "react";
import { createIntl, RawIntlProvider } from "react-intl";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider, createTheme } from "@mui/material/styles";

import { LANG_CODES } from "./common/constants";
import { appMessages } from "./lang";

const theme = createTheme({
  palette: {},
});
const intl = createIntl({
  locale: LANG_CODES.EN,
  messages: appMessages(LANG_CODES.EN),
});

const AllTheProviders: ComponentType<PropsWithChildren> = ({
  children,
}): any => (
  <ThemeProvider theme={theme}>
    <RawIntlProvider value={intl}>{children}</RawIntlProvider>
  </ThemeProvider>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";

const mockAppStore = {
  spreadsheet: {
    data: [
      [
        {
          id: "A0",
          value: "3",
          formula: "",
        },
        {
          id: "B0",
          value: "18",
          formula: "",
        },
        {
          id: "C0",
          value: 13.5,
          formula: "=12 + 3 / (2)",
        },
      ],
    ],
    byId: {
      A0: [0, 0],
      B0: [0, 1],
      C0: [0, 2],
    },
    progress: {},
    loading: false,
    error: false,
  },
};

export { customRender as render, mockAppStore };
