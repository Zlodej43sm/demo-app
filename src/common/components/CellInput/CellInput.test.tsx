import React from "react";
import { useDispatch, useSelector } from "react-redux";

import { render, mockAppStore } from "../../../test-utils";
import CellInput from ".";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe("CellInput", () => {
  const props = {
    id: "A0",
    setCurrentEditRow: jest.fn(),
  };
  const mockedUseSelector = useSelector as jest.Mock<typeof useSelector>;
  const mockedDispatch = useDispatch as jest.Mock<typeof useDispatch>;
  const mockStore = {
    ...mockAppStore,
  };

  beforeEach(() => {
    mockedDispatch.mockReturnValue(jest.fn());
    mockedUseSelector.mockImplementation((selector) => selector(mockStore));
  });

  it("SHOULD match snapshot", () => {
    const { container } = render(<CellInput {...props} />);
    expect(container).toMatchSnapshot();
  });
});
