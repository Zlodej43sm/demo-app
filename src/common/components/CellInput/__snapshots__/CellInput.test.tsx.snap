// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellInput SHOULD match snapshot 1`] = `
<div>
  <div
    class="Spreadsheet-editBoxRoot-1 MuiBox-root css-0"
  >
    3
    <svg
      class="Spreadsheet-editIcon-2"
    >
      edit.svg
    </svg>
  </div>
</div>
`;
